### PR TITLE
Add Polygon Type To Schema / PolygonContain to Query

### PR DESCRIPTION
**Pull request by dplewis**
https://github.com/parse-community/parse-server/issues/3942

Adds Mongo and PG support for Polygon Type and PolygonContain query

### CODE REVIEW

**Comment by dplewis** [comment]
@flovilmart Done, Let me know if I need more tests

**Comment by dplewis** [comment]
I found that mongo throws an error for self-intersecting polygons. Is it ok if mongo just throws an error?

`Loop is not valid`

https://github.com/mongodb/mongo/blob/master/src/mongo/db/geo/geoparser.cpp#L223

**Comment by flovilmart** [comment]
Is it possible to enforce the indexing when adding those columns types the same way we do with 2D indexes?

**Comment by dplewis** [comment]
Since $geoIntersect doesn't require a index I can't catch an error to add one like $near.

I could check the schema and if Polygon Type add an index.

I'm going to write a few more test for postgres. It doesn't throw errors right now but It may for indexes.

**Comment by dplewis** [comment]
@flovilmart how does this look?

**Comment by flovilmart** [comment]
I’ll have a deeper look after the 2.5.0 release :)

**Comment by flovilmart** [comment]
Why the removal of the test,

**Comment by dplewis** [comment]
I felt like it wasn't needed since I'm not checking for invalid loops. I can add it back if it helps.
I noticed when a mongoError gets thrown it retries the request a few times. Is that normal?

**Comment by flovilmart** [comment]
> noticed when a mongoError gets thrown it retries the request a few times.

The JS SDK is retrying dumbly.

File: src/Adapters/Storage/Mongo/MongoSchemaCollection.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ function mongoFieldToParseSchemaField(type) {
   case 'geopoint': return {type: 'GeoPoint'};
   case 'file':     return {type: 'File'};
   case 'bytes':    return {type: 'Bytes'};
+  case 'polygon':    return {type: 'Polygon'};
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoSchemaCollection.js]: nit alignment

File: src/Adapters/Storage/Mongo/MongoSchemaCollection.js [hunk 2]
```diff
@@ -98,6 +99,7 @@ function parseFieldTypeToMongoFieldType({ type, targetClass }) {
   case 'GeoPoint': return 'geopoint';
   case 'File':     return 'file';
   case 'Bytes':    return 'bytes';
+  case 'Polygon':    return 'polygon';
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoSchemaCollection.js]: nit alignment

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 3]
```diff
@@ -182,7 +182,8 @@ export class MongoStorageAdapter {
 
   addFieldIfNotExists(className, fieldName, type) {
     return this._schemaCollection()
-      .then(schemaCollection => schemaCollection.addFieldIfNotExists(className, fieldName, type));
+      .then(schemaCollection => schemaCollection.addFieldIfNotExists(className, fieldName, type))
+      .then(() => this.create2dsphereIndex(className, fieldName, type));
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: can you rename `create2dsphereIndex` into an umbrella `createIndexesIfNeeded`? more explicit :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 4]
```diff
@@ -429,6 +430,20 @@ export class MongoStorageAdapter {
     return this._adaptiveCollection(className)
       .then(collection => collection._mongoCollection.createIndex(index));
   }
+
+  create2dsphereIndex(className, fieldName, type) {
+    if (type && type.type === 'Polygon') {
+      const index = {}
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: you can use:

```
const index = {
  [fieldName]: '2dsphere'
};
```

and remove the next line

Comment (dplewis) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Works nicely. Thanks

**Comment by dplewis** [comment]
Thanks for the feedback. Changes have been made

**Pull request merged**